pytest 5.4+ doesn't handle IsolatedAsyncioTestCase from CPython core.

So, https://github.com/pytest-dev/pytest/pull/5734 causes pytest to incorrectly skip some async tests, specifically those based on https://docs.python.org/3/library/unittest.html#unittest.IsolatedAsyncioTestCase or code like it

See https://github.com/pytest-dev/pytest/pull/5734#issuecomment-599023047 for real world failure, the mock backport has some of its own code in this area, to cater for Python 3.6 and 3.7:
https://github.com/testing-cabal/mock/blob/b5ce0a5c4d372b77deff46fec8edf974d7d1f875/mock/backports.py#L38-L82

For now, I'll just pin to pytest<5.4, but would be interested to know in how this will be fixed.

```
$ pytest mock/tests/test_reproducer.py 
========================================================= test session starts =========================================================
platform darwin -- Python 3.8.1, pytest-5.3.5, py-1.8.1, pluggy-0.13.1
rootdir: /Users/chris/vcs/git/mock, inifile: setup.cfg
plugins: cov-2.8.1
collected 1 item                                                                                                                      

mock/tests/test_reproducer.py F                                                                                                 [100%]

============================================================== FAILURES ===============================================================
_________________________________________________ AsyncArguments.test_something_async _________________________________________________

self = <mock.tests.test_reproducer.AsyncArguments testMethod=test_something_async>

    async def test_something_async(self):
    
        async def addition(x, y):
            return x+y
    
>       self.assertEqual(await addition(2, 2), 3)
E       AssertionError: 4 != 3

mock/tests/test_reproducer.py:11: AssertionError
```

IsolatedAsyncioTestCase's run method does all the async stuff, in the same way an async plugin would:

```
        def run(self, result=None):
            self._setupAsyncioLoop()
            try:
                return super().run(result)
            finally:
                self._tearDownAsyncioLoop()
```

Remember: this is a unittest.TestCase subclass, so it can do lots in that run method...

--- a/changelog/6924.bugfix.rst
+++ b/changelog/6924.bugfix.rst
@@ -1,0 +1,1 @@
+Ensure a ``unittest.IsolatedAsyncioTestCase`` is actually awaited.

--- a/src/_pytest/python.py
+++ b/src/_pytest/python.py
@@ -175,15 +175,33 @@ def async_warn(nodeid: str) -> None:
 @hookimpl(trylast=True)
 def pytest_pyfunc_call(pyfuncitem: "Function"):
     testfunction = pyfuncitem.obj
-    if iscoroutinefunction(testfunction) or (
-        sys.version_info >= (3, 6) and inspect.isasyncgenfunction(testfunction)
-    ):
+
+    try:
+        # ignoring type as the import is  invalid in py37  and mypy thinks its a error
+        from unittest import IsolatedAsyncioTestCase  # type: ignore
+    except ImportError:
+        async_ok_in_stdlib = False
+    else:
+        async_ok_in_stdlib = isinstance(
+            getattr(testfunction, "__self__", None), IsolatedAsyncioTestCase
+        )
+
+    if (
+        iscoroutinefunction(testfunction)
+        or (sys.version_info >= (3, 6) and inspect.isasyncgenfunction(testfunction))
+    ) and not async_ok_in_stdlib:
         async_warn(pyfuncitem.nodeid)
     funcargs = pyfuncitem.funcargs
     testargs = {arg: funcargs[arg] for arg in pyfuncitem._fixtureinfo.argnames}
     result = testfunction(**testargs)
     if hasattr(result, "__await__") or hasattr(result, "__aiter__"):
-        async_warn(pyfuncitem.nodeid)
+        if async_ok_in_stdlib:
+            # todo: investigate moving this to the unittest plugin
+            #       by a test call result hook
+            testcase = testfunction.__self__
+            testcase._callMaybeAsync(lambda: result)
+        else:
+            async_warn(pyfuncitem.nodeid)
     return True
 
 

--- a/testing/example_scripts/pytest.ini
+++ b/testing/example_scripts/pytest.ini
@@ -1,0 +1,2 @@
+[pytest]
+# dummy pytest.ini to ease direct running of example scripts

--- a/testing/example_scripts/unittest/test_unittest_asyncio.py
+++ b/testing/example_scripts/unittest/test_unittest_asyncio.py
@@ -1,0 +1,15 @@
+from unittest import IsolatedAsyncioTestCase  # type: ignore
+
+
+class AsyncArguments(IsolatedAsyncioTestCase):
+    async def test_something_async(self):
+        async def addition(x, y):
+            return x + y
+
+        self.assertEqual(await addition(2, 2), 4)
+
+    async def test_something_async_fails(self):
+        async def addition(x, y):
+            return x + y
+
+        self.assertEqual(await addition(2, 2), 3)

--- a/testing/test_unittest.py
+++ b/testing/test_unittest.py
@@ -1129,3 +1129,11 @@ def test_trace(testdir, monkeypatch):
     result = testdir.runpytest("--trace", str(p1))
     assert len(calls) == 2
     assert result.ret == 0
+
+
+def test_async_support(testdir):
+    pytest.importorskip("unittest.async_case")
+
+    testdir.copy_example("unittest/test_unittest_asyncio.py")
+    reprec = testdir.inline_run()
+    reprec.assertoutcome(failed=1, passed=1)

